### DISCUSSION TITLE
`allow-unwrap-in-tests` vs `#![cfg_attr(test, allow(clippy::unwrap_used))]`

### DISCUSSION THREAD

**Discussion by mladedav** [rust-lang/rust-clippy#13779] [Q&A]
I am wondering why does the option exist in clippy.toml. Is there a difference between these two?

**Comment by flip1995** [comment]
Using `cfg_attr(test)` will disable the lint in all code, not only test code when you run `cargo clippy --tests` or `cargo clippy --all-targets`. The setting in the `clippy.toml` file will make the lint check, if the `unwrap` is in a test context (a function marked with `#[test]` or a test module), and not trigger there.

**Reply by mladedav** [reply]
Having this code still emits an error for the first unwrap while it does not for the second one even with `cargo clippy --all-targets` and `cargo clippy --tests`:

```rust
#![allow(clippy::unnecessary_literal_unwrap)]
#![allow(dead_code)]
#![warn(clippy::unwrap_used)]
#![cfg_attr(test, allow(clippy::unwrap_used))]

fn code() {
    Some("code").unwrap();
}

#[test]
fn test() {
    Some("test").unwrap();
}
```

If I understand correctly your answer, this example should not have emitted anything, right?

**Reply by mladedav** [reply]
Gentle ping @flip1995. I just want to know if I maybe misunderstood your answer or there is something else happening during the compilation.

I get this is a minor thing so if you don't have the time to dig into the details about this, that's fine too.

**Reply by mladedav** [reply]
So why this works is (based on [this issue](https://github.com/rust-lang/cargo/issues/851)) that the lib is built once normally to be linked for doc tests and then is built once with `--cfg test` for actual tests which produces a binary with test harness.

So it works although it is because of cargo internals that might not be considered stable but it's not going to change in the foreseeable future.
